### ISSUE TITLE
IPython 8 regression: escaping insert mode in vi editing mode takes too long.

### ISSUE THREAD

**Issue by mario-grgic** [ipython/ipython#13443]
After updating to IPython 8.0 and with vi editing mode enabled, it now takes about 1 second to exit insert mode and go back to nav mode. Even worse, if you hit ESC (to exit insert mode) and continue typing your editing command, exiting inset mode is cancelled altogether. 

This is extremely jarring. For example: type something in insert mode hit ESC and start typing editing commands (e.g. dd to delete a line) and instead you get single d echoed in the line and exiting insert mode is cancelled altogether (you stay in insert mode).

This is a regression introduced in IPython 8, and there doesn't appear to be a setting that controls this delay.

**Comment by mario-grgic** [comment]
It appears that if you flip the setting 

`c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False`

 this is fixed. So, some Emacs bindings must use Escape + <some_other_key> sequence, which means we have to wait after ESC for another key press.

Perhaps solution here is to not add Emacs bindings by default if vi editing mode is used?

**Comment by Ellana42** [comment]
Same issue here. The setting helped with the editing commands not being written, but leaving insert mode is still slow and the commands I issue while switching from insert mode to normal mode are not taken into account. For example a Esc + dd typed fast will not delete the whole line.

**Comment by 0xConnorRhodes** [comment]
I also have the same issue as Ellana42. Adding `c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False` does not get rid of the lag when switching to normal mode. Notably switching from normal mode to inset mode is instant. It's only switching from insert to normal mode which is slow.

**Comment by jerpint** [comment]
Same issue as well, it makes vi mode essentially unusable

**Comment by willthbill** [comment]
The following configuration in `~/.ipython/profile_default/ipython_config.py`, makes it much more comfortable to work with.
```
c.TerminalInteractiveShell.editing_mode = 'vi'
c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False
c.TerminalInteractiveShell.timeoutlen = 0.25
```
`c.TerminalInteractiveShell.timeoutlen` has a default value of 0.5. Setting it to 0 makes <esc> from insert mode instant, but it also makes `dd` not work properly. Maybe some kind of setting to disable delays in insert mode would solve this completely?

**Comment by g15ecb** [comment]
In addition to that mentioned above I also noticed that the cursor shape is set to `|` in the terminal upon exit of IPython 8 if you exit when in 'insert mode'; exiting in `normal mode` will leave the terminal default cursor shape intact. I am using zsh and have `bindkey -v` in `~/.zshrc`.

**Comment by jtherrmann** [comment]
I'm having the same issue with `7.31.1`, does anyone at least have a solution for that version?

**Comment by MathieuTuli** [comment]
this took me a while to figure out and plagued me for so long! having emacs bindings default off if using vim seems like a good idea to me. I spent a while thinking the vi-mode was just not working

**Comment by max-sixty** [comment]
The `c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False` works for me!

One disadvantage of this is that `C-w` is useful, and that's turned off. If anyone figures out how to get the previous behavior back, please do share!

**Comment by ianbrody** [comment]
Maybe this should be a different issue, but it's also treating `f<letter>` in normal mode as subject to the timeout. I should be able to type `f` in normal mode, wait a long time, and hit a letter, and have it jump to that letter.

**Comment by bbliem** [comment]
> One disadvantage of this is that `C-w` is useful, and that's turned off. If anyone figures out how to get the previous behavior back, please do share!

@max-sixty I managed to get CTRL-W to work by putting the following in `~/.ipython/profile_default/startup/00-keybindings.py` (adapted from the documentation about [keyboard shortcuts](https://ipython.readthedocs.io/en/latest/config/details.html#keyboard-shortcuts))

```
from IPython import get_ipython
from prompt_toolkit.enums import DEFAULT_BUFFER
from prompt_toolkit.keys import Keys
from prompt_toolkit.filters import HasFocus, HasSelection, ViInsertMode, EmacsInsertMode

from prompt_toolkit.key_binding.bindings.named_commands import get_by_name

ip = get_ipython()
insert_mode = ViInsertMode() | EmacsInsertMode()

# Register the shortcut if IPython is using prompt_toolkit
if getattr(ip, 'pt_app', None):
    registry = ip.pt_app.key_bindings
    registry.add_binding(Keys.ControlW,
                     filter=(HasFocus(DEFAULT_BUFFER)
                             & ~HasSelection()
                             & insert_mode))(get_by_name('backward-kill-word'))
```

My ipython config file contains this, by the way:
```
c.TerminalInteractiveShell.editing_mode = 'vi'
c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False
```

**Comment by jerpint** [comment]
@minjae your solution of commenting out seems to work on my end, thanks!

**Comment by jacktose** [comment]
@minjae Thanks for that deep info! Did you ever open a PR to straighten out the behavior?

**Comment by jtherrmann** [comment]
I opened https://github.com/ipython/ipython/pull/13861 for @minjae's workaround to get this moving forward. I'm not familiar with the ipython codebase so don't know if this is the best solution.

**Comment by minjae** [comment]
@jtherrmann thanks for getting to making the PR.  I was initially hoping for more concrete feedback, but even months after, it seems like there isn't much feedback.  I still do not know what the correct behavior of `reformat_and_execute` is.  At least on my machine, it does not seem to reformat anything (but it executes).  Could someone chime in before we go ahead with the PR?

**Comment by jtherrmann** [comment]
I confirmed that this is now fixed as of IPython 8.8.0. To summarize, your config file must have the following lines:

```
c.TerminalInteractiveShell.editing_mode = 'vi'
c.TerminalInteractiveShell.emacs_bindings_in_vi_insert_mode = False
```

No other workarounds are required.

The second line is required for eliminating the delay when escaping insert mode. Thanks again to @minjae for providing the workaround in https://github.com/ipython/ipython/issues/13443#issuecomment-1172097730.
